### ISSUE TITLE
Support Namespaced Managed Resources

### ISSUE THREAD

**Issue by ctrox** [crossplane/crossplane-runtime#251]
### What problem are you facing?

Right now, crossplane-runtime does not formally support namespaced managed resources. It kind of works and only breaks if certain features are used. Because we (my team and I) were unaware that this was unintentional, we simply started using it.

### How could Crossplane help solve your problem?

It would be great if crossplane-runtime formally supported namespaced managed resources. I initially thought this [was a bug](https://github.com/crossplane/crossplane-runtime/pull/246), but it turns out this was simply not intended to work. But most of crossplane-runtime already works when using a namespaced managed resource, just by instructing kubebuilder: `+kubebuilder:resource:scope=Namespaced`. While we are aware of all the other Crossplane constructs that allow to have namespaced resources (compositions, publications etc.), it introduces a lot of configuration overhead, where we just could use the managed resource directly for some resources.

After a few glances at the code, it seems only a few changes here and there would be required to make references, externalName etc. work properly with namespaced resources. If you are not totally opposed to this, I can offer to contribute the required changes.

**Comment by nekto0n** [comment]
Hey there! One possible user story could go like this:
>I am on the infrastructure team at my company and we want to offer Azure (AKS in particular) as a service for engineers via crossplane controller. We want to be able to apply policies, grant permissions, attach quotas and perform cost and audit control.

It seems that logical to achieve this by using namespaces to get k8s native tools to work (RBAC, etc). Does that sound like a legit case?

**Comment by negz** [comment]
> I am on the infrastructure team at my company and we want to offer Azure (AKS in particular) as a service for engineers via crossplane controller. We want to be able to apply policies, grant permissions, attach quotas and perform cost and audit control.

Crossplane is designed to support this via resource claims (which are namespaced). The idea is that you'd offer AKS as a service via a claim, not via the 'raw' managed resources.

> For our usecase namespaced managed resources would allow us to easily delete all resources created for a deployment. We create a namespace per application. To remove an application we simply delete the namespace and all resources for that application are gone. With non-namespaced crossplane resources we must also label the resources and must know which resources are created by the application.

FWIW, I believe this is also something you could do with claims. Deleting a namespace containing Crossplane composite resource claims (XRCs) would cause the claims to be deleted, which in turn would delete the XR and all composed resources.

Generally, I don't think we'll be adding support for namespaced managed resources - at least not without a major version bump. I will leave this issue open to gather more feedback.

**Comment by nekto0n** [comment]
> Crossplane is designed to support this via resource claims (which are namespaced). The idea is that you'd offer AKS as a service via a claim, not via the 'raw' managed resources.

Got it! Please correct me if I got the idea wrong: if infrastructure owners want to expose all available AWS resources as claims, they need to manually create corresponding claim CRDs. Are there any means to automate this procedure?

**Comment by luebken** [comment]
Mr @benagricola has created a generator https://github.com/benagricola/crossplane-composition-generator

It's on my to do list to give it a try.

**Comment by JohnLBevan** [comment]
Allegedly this is coming in Crossplane 2.0 (see 27:10 of https://youtu.be/8zQtpcxmdhs?t=1632)

**Comment by jbw976** [comment]
Yes, namespaced managed resources are supported in the v2 preview: https://github.com/crossplane/crossplane/releases/tag/v2.0.0-preview.1 🎉
